### ISSUE TITLE
How to embed vector based images...

### ISSUE THREAD

**Issue by markwjkelly** [rk700/PyMuPDF#118]
I have been experimenting with the system with great results.  I am looking for an option to embed a vector based image rather than pixel / raster image.  Is this possible?  I have been reviewing the docs on "readthedocs.io" but having found any info or examples.  Some pointers would be perfect. THX

**Comment by JorjMcKie** [comment]
I am glad you find PyMuPDF useful.

Could you please provide an example for the type of graphics you have in mind?

In principle, we are limited to what MuPDF supports in that area (which in turn is a subset of what PDF itself supports, though a large one). But we do not yet support the whole set of MuPDF's features.

We currently can import SVG images, see [here](https://pymupdf.readthedocs.io/en/latest/pixmap/#imagefiles).

**Comment by markwjkelly** [comment]
I tried the SVG options (filed exported from Illustrator using SVG 1.0 standard) but get unknown imge file format error. I have attached the image file and the same code I am using ...
`import fitz                          # <-- PyMuPDF
doc = fitz.open("calender.pdf")          # open the PDF
rect = fitz.Rect(700, 0, 820, 18)     # where to put image: use upper left corner
pix = fitz.Pixmap("silicon2018.svg")
for page in doc:
    page.insertImage(rect, pixmap = pix)
#    page.insertImage(rect, filename = "silicon2018.svg")

doc.saveIncr() `
Here is a link to the SVG file as Github wont let me attach it...
http://siliconstorm.com/silicon2018.svg

Any suggestions?  Thanks again for your help.

**Comment by JorjMcKie** [comment]
You are right - getting the same error.
What - confusingly - **does** work is the following:
```
>>> pix_doc = fitz.open("silicon2018.svg")
>>> pix = pix_doc.getPagePixmap(0)
>>> pix
fitz.Pixmap(DeviceRGB, fitz.IRect(0, 0, 470, 72), 1)
>>> 
```

The pixmap then can be used as any other image in `Page.insertImage` methods.
So:
Like all other image files, SVG also is one of the supported (Py) MuPDF **document** types. Which can be opened as 1-page documents, and the page of which can be rendered as a pixmap.
When rendering the page of such documents as a pixmap, all options for pixmap creation are available, especially scaling, rotating, ... transparency setting. The above example lets alpha default to transparency. The statement `pix = pix_doc.getPagePixmap(0, alpha = 0)` would prevent creating an alpha channel.

I will submit an inquiry to MuPDF whether they are aware of this and if this is intentional.

**Comment by JorjMcKie** [comment]
I think I have an explanation:
SVG files are an XML subtype, and MuPDF supports XML and (X)HTML files as document types - not as images. So SVG images are only available via the detour above, as rendered pages.

**Comment by markwjkelly** [comment]
That makes sense.  Unfortunately, it still adds the image as a rendered object.  I am trying to get the vector version into the document so that it looks good when scaled. 
Thank you again for looking into the issue.

**Comment by JorjMcKie** [comment]
Eah, I know. That's unfortunate, but I know no way around that one.
The only alleviation I can offer is using scale/zoom matrix when creating the pixmap. This at least postpone the point when you start seeing the pixels in zooming the result PDF.
But of course the created (raster) image will also be proportionally larger that gets stored. Of course (as a solace) it will only be stored once per PDF ...
In case of using the defaulting Identity matrix the resulting PNG image has a size of 10KB, if you use fitz.Matrix(2,2) (zoom factor 2) the result has only a little more than the double size: 23KB.
If you use `garbage=4,  deflate=True` when saving the PDF this should be your effect, too.

**Comment by JorjMcKie** [comment]
Based on your input I have started investigating options for fullfilling your requirement. Currently I am focussing on two (similar) alternatives.
Both of them are based on not directly displaying SVG images, but instead PDF versions of SVGs. I still have not found a way to directly store (and get displayed) SVG images in a PDF - independently of MuPDF (maybe you know a way?).
On the other hand, there exists a plethora of SVG-to-PDF tools out there. I used the Python tool `svglib` to convert your `silicon2018.svg` into `silicon2018.pdf`. The main trick is the conversion of SVG syntax into equivalent PDF operator syntax. I have noticed others like `Apache Batik` and `Apache FOP` doing more or less the same thing (didn't try them because they are Java based, which I don't have and don't want to have).

Anyway - the 2 alternatives start from here.

It looks feasible to display the image of a **source** PDF page (like the above `silicon2018.pdf`, which could look similar to this: [tsvg.pdf](https://github.com/rk700/PyMuPDF/files/1604013/tsvg.pdf))  in a **containing** PDF page (in **vector** form, not as a raster image).

The attached example stands for the **first alternative**:
Basically, a source PDF page is copied over to a target PDF as an `XObject`. The image of that `XObject` can be displayed several times on the same or other pages of the target.
The challenge with this solution is that, in order to support general source PDF pages (not only SVG conversions), a lot of other, nested objects referenced by the source page need to be copied over (and correctly link together) as well,  like images, fonts, annotations, ...

An alternatives seems to be embedding the complete source PDF in the target and reference its pages via a special subtype of `XObjects`. I am going to investigate this as well now.

I would welcome any feedback you may have ...

**Comment by JorjMcKie** [comment]
I just uploaded changes to the current v1.12.1 which add some vector image capability as announced in my previous comment.
I ended up choosing my first alternative: you can display arbitrary PDF pages in selected rectangles of the containing page(s). They should display exactly like the original on any zooming level.
Look into the Wikis to find an example for creating 4up-versions of input PDFs.
I hope this serves your requirements.

Creating SVG output of a page is now also possible via `getSVGimage` method.

**Comment by JorjMcKie** [comment]
Just as a proper completion, here is a script that puts an SVG image as a logo on each page of a PDF (as a **vector image** of course). Use as `python svglogo.py input.pdf image.svg`. In order for this to work, you need the reportlab and svglib packages:

```
import fitz
from svglib.svglib import svg2rlg
from reportlab.graphics import renderPDF
import sys
doc_fn = sys.argv[1]                        # name of PDF file
svg_fn = sys.argv[2]                        # name of SVG image file

drawing = svg2rlg(svg_fn)                   # open the SVG
pdfbytes = renderPDF.drawToString(drawing)  # convert SVG to PDF stream
src = fitz.open("pdf", pdfbytes)            # open PDF-converted SVG
doc = fitz.open(doc_fn)                     # open PDF to be modified
rect = fitz.Rect(0, 0, 50, 50)              # logo rectangle on each page
for page in doc:                            # scan through PDF pages
    page.showPDFpage(rect, src, 0)          # put logo on page
  
doc.save("logo-" + doc_fn)                  # save PDF as a new name

```
Method `showPDFpage` offers the option `keep_proportions` (default `True`) to allow / suppress independent zooming of the edges, and the option `overlay` (default `True` = foreground) to control image positioning.

**Comment by markwjkelly** [comment]
I tried the latest version with your example but it seems to be throwing an error...
'Page' object has no attribute 'showPDFpage'.  
Will keep trying some alternative options
... and thank you for following up on this functionality.

**Comment by JorjMcKie** [comment]
Hm, what is your version date? Maybe download again ... but read to the end before you do.
Sorry, I feel a little embarrassed for not having created a separate new version (1.12.x) for it ... But this is such a hassle for my 10 Python / bitness combinations.

**Second comment:**
I realized that garbage collection (done by underlying C-library MuPDF) does **not recognize** multiple copies of this type of image (PDF type `XObject`). So if you put a logo on 1000 pages, the resulting PDF will contain 1000 copies of it ... unbearable. So I have been busy developing a way out over the last 12 hours. It works like this:

```
import fitz
from svglib.svglib import svg2rlg
from reportlab.graphics import renderPDF
import sys
doc_fn = sys.argv[1]                        # name of PDF file
svg_fn = sys.argv[2]                        # name of SVG image file

drawing = svg2rlg(svg_fn)                   # open the SVG
pdfbytes = renderPDF.drawToString(drawing)  # convert SVG to PDF stream
src = fitz.open("pdf", pdfbytes)            # open PDF-converted SVG
doc = fitz.open(doc_fn)                     # open PDF to be modified
rect = fitz.Rect(0, 0, 50, 50)              # logo rectangle on each page
xref = 0                                    # xref of stored image
for page in doc:                            # scan through PDF pages
    xref = page.showPDFpage(rect, src, 0,   # put logo on page
                    reuse_xref = xref)      # xref>0 prevents multiple copies
  
doc.save("logo-" + doc_fn)                  # save PDF as a new name
```

I am finishing tests. Please wait for my message before you download again.

**Comment by JorjMcKie** [comment]
BTW - failure to garbage collect multiple XObjects is a bug IMHO. I filed it at MuPDF's site. But I don't believe we will see a solution soon.

**Comment by JorjMcKie** [comment]
Just updloaded a new version to this repo and also to [optional materials](https://github.com/JorjMcKie/PyMuPDF-Optional-Material) and [wheels](https://github.com/JorjMcKie/PyMuPDF-wheels).

Good luck.

**Comment by markwjkelly** [comment]
I have had several successes with the new functionality - Thanks you.  I am working through a batch of tests for a variety of SVG logos (Color, dimensions, effects, gradients, etc)  I have found a few cases that are not supported (Linear and Radial gradients) and some curved edges.  For example, the Silicon Storm that I was initially using for test only comes through and black on white (No color on  the ring) Other SVG logos show full color. I haven't tried digging deeper yet but plan to over the next few days. Happy to share source files and findings with all.

**Comment by JorjMcKie** [comment]
Thanks for your feedback!
Yes, I have seen, too, that `silicon2018.svg` is b/w only.  When I looked at this `svglib` package, I also read, that they actively are declaring limitations of what they are able to crunch.

Luckily for PyMuPDF, this is outside our scope :-) svglib, after all, is just an example for such funtionality.
Or more seriously said: maybe there are better, more complete converters out there, like the aforementioned Apache Batik, etc.

**Comment by JorjMcKie** [comment]
I am going to close this issue again now.

Just as an announcement: I have modified `Page.showPDFpage()` somewhat again - to be uploaded soon.
Without changes to the API, it will now have a significantly better performance, which is mostly noticable in cases other than yours. For example, to combine several PDF input pages in one output page now runs twice as fast and produces an 80+% smaller output ...
Numbers for the Adobe manual (1'310 pages, 31 MB, each 4 input pages producing 1 output page) before were 5 sec and 92 MB output. Now I have 2 sec and 11 MB output.

**Comment by stefaneidelloth** [comment]
Thank you. The above exaples helped me to include some *.svg figures as vector graphcis in pdf. 

* `page.showPDFpage` is now `page_show_pdf_page`. 
* Interactivity of svg files is not supported.
